scipy/stats/_wilcoxon.py

```diff
@@ -197,18 +206,26 @@ def _wilcoxon_nd(x, y=None, zero_method='wilcox', correction=True,
                  alternative='two-sided', method='auto', axis=0):
 
     temp = _wilcoxon_iv(x, y, zero_method, correction, alternative, method, axis)
-    d, zero_method, correction, alternative, method, axis, output_z = temp
+    d, zero_method, correction, alternative, method, axis, output_z, m_auto = temp
 
     if d.size == 0:
         NaN = _get_nan(d)
         res = _morestats.WilcoxonResult(statistic=NaN, pvalue=NaN)
-        if method == 'approx':
+        if method == 'asymptotic':
             res.zstatistic = NaN
         return res
 
-    r_plus, r_minus, se, z, count = _wilcoxon_statistic(d, zero_method)
+    r_plus, r_minus, se, z, count, has_ties = _wilcoxon_statistic(d, zero_method)
+
+    # we only know if there are ties after computing the statistic and not
+    # at the input validation stage. if the original method was auto and
+    # the decision was to use an exact test, we override this to
+    # a permutation test now (since method=`exact`` is not exact in the
+    # presence of ties)
+    if m_auto and has_ties and method == 'exact':
```

Comment:
```suggestion
    if (has_zeros or has_ties) and method == 'auto':
```
Why not leave the `method` unresolved until here (instead of changing it and changing it again based on a variable with a name that is less obvious than `has_zeros`)?

Reply:
true, that is better. i made a suggestion that resolves to exact or PermutationMethod later.